System Tests

justintaft opened the pull request:
~~DO NOT MERGE. EXPERIMENTAL.~~

Whenever light table runs remote debugging is available. This is a security concern and we may want to change this in the future...
https://github.com/LightTable/LightTable/blob/baa2037aecaf382368ef299ed885414ca0f8e1ac/deploy/core/main.js#L107

To access remote debugging tools:
-  run light table as normal
-  navigate to http://localhost:8315

To debug with web driver:
- run chromedriver -v
- run the  contents of system-tests.clj (binary location will need to be changed)
- Launch light table

Sometimes light table does not initialize all the way. If this happens, close light table & retry.

I would like the system tests to use a cljs or js library, so we don't have to boot up the jvm for testing. Any Thoughts?

Commit:
Adds system tests w/ travis and appveyor ci intergration

kenny-evitt commented:
Thanks! This will be great to get working.

justintaft commented:
Not sure how to use send-keys properly with taxi with switch-to-active call...

justintaft commented:
Hm, made a issue ticket on clj-webdriver for discussing semantics of taxi/switch-to-active. IMO this probably won't be used too often. We can create a simple function called "current-element" which can be used as the query value for taxi if need be. 

https://github.com/semperos/clj-webdriver/issues/153

kenny-evitt commented:
@justintaft FYI, let's use this PR for ongoing discussion.

justintaft commented:
There is a command "Dev: Open Developer Tools". This shows the familiar developer tools in chrome. Easy way to inspect the DOM =).

We should have a system test for creating, saving, and loading a file. However, a native file dialog window pops up during these processes and you can't use web driver to click save. In other projects I've wrote a program to click native buttons, but it was always a hacky solution.... anyone have a better suggestion? This file dialog window also appears when adding a file/directory to the workspace tree.

For now I can keep things simple.  I'll create a temporary file, launch the editor passing in the path to this file, verify the contents are correct, modify the contents, save, then verify the changes have been stored to disk. Since the file exists, we won't have the dialog window pop up for now...

kenny-evitt commented:
@justintaft Would it overly defeat the purposes of these tests to fake or mock the native OS UX? We don't need to test it directly – it should be enough that the fake or mock returns specific values to the calling code.

justintaft commented:
This should be ok. 

If not done so already,  their should exist a cljs  function used to show the dialog window.  This will contain the electron call to show the native file window selection screen.  We can then assert this code is correct in a test,  and then replace it with our mocking code. 

I was thinking about injecting a simple form that has a text  field for a file path and a OK button. That way our system tests still can read nice and the used file paths are not hidden.

justintaft commented:
@kenny-evitt 

Needs a bit of refactoring, but the dialog window should be mockable now =).

kenny-evitt commented:
@justintaft Nice!

kenny-evitt commented:
@justintaft We have [a QA checklist of tests](https://github.com/LightTable/LightTable/wiki/QA-Checklist) to do manually prior to release – these would be great to automate.

justintaft commented:
@kenny-evitt Thanks =) 

Hopefully once I get everything bootstrapped I can get some  help from the community.

Ok, so I wrote the code for the mock dialog window. Question is, how do we inject this mock?

Few options:
-Write a "test" plugin which is used to setup the environment. 
-Compile the script and modify the user behavior to autoload the js.
-Open up a repl, type the code in, and evaluate (this is really nasty).

Is there a way to sandbox light table instance so it won't find my personal user behavior file & installed plugins?

justintaft commented:
Ah alright, looks like the "user" directory is overridable.

https://github.com/LightTable/LightTable/blob/1d9dd0cc3154048834aedd554a056acb0a2df127/src/lt/objs/files.cljs#L360

As for injecting the mock, I'll go the plugin route for now, unless otherwise suggested =).

kenny-evitt commented:
@justintaft A plugin seems perfectly fine.

justintaft commented:
@kenny-evitt 

Added some assertions for saving & loading a file.

 I couldn't use  `$('.CoreMirror-code')[0].CodeMirror` to get an instance. Webdriver reported `$` wasn't defined. Weird thing is in dev tools, `$` is available... I just asserted against the text in the editor for now.

justintaft commented:
@kenny-evitt 

*Save any local work before using scripts below*

script/run-tests.sh should run the automated tests.

It downloads & starts chromedriver, and then runs `lein test`. The run test script needs to handle different operating systems. I updated travis yml to auto run them, but have not tested everything. 

 If you need to download the mac version of chrome driver, then uncomment the mac line in run-tests.sh.

Lastly, light table doesn't initialize all the way for whatever reason when chrome driver is running. Running "lein test" a few times usually gets the tests running. You might have to kill the old light table process if it hangs.

kenny-evitt commented:
@justintaft This is really great work. :clap:

I'm working on manually QA-ing the 0.8.0 alpha release now but I'd like to try to cover those tests with this stuff.

justintaft commented:
@kenny-evitt 

![Reaction](http://f.cl.ly/items/3C121F3E0P0A2o092I0w/output_optimized.gif)
https://media.giphy.com/media/GA2FNpP1kAQNi/giphy.gif

It looks likes the tests are running on travis =). I wanted 1 to fail, other 2 should succeed. Was a quick test to make sure things are set up right.

https://travis-ci.org/LightTable/LightTable/builds/83830956#L443

justintaft commented:
Found a cygwin appveyor.yml project example
https://github.com/DynamicDevices/mono/blob/master/appveyor.yml

justintaft commented:
@kenny-evitt 

Windows CI setup is coming along. I still need to download chrome for windows & unzip in the right place. The console output is looking good though. 

https://ci.appveyor.com/project/kenny-evitt/lighttable/build/1.0.36

kenny-evitt commented:
@justintaft looking good

justintaft commented:
@kenny-evitt   tests are running on windows AND linux now =)

Again, the failure is expected
https://travis-ci.org/LightTable/LightTable/builds/86075536#L443
https://ci.appveyor.com/project/kenny-evitt/lighttable/build/1.0.81#L451

kenny-evitt commented:
@justintaft Thanks for working on this. What's next?

justintaft commented:
The following files launch light table while capturing output and redirecting to a file...
this seems a bit hacky and clutters up the script directory...

script/light-test.bat
script/light-test.sh

I have to research how to run & kill a light-table process. Currently using shell/sh from clojure library, but output is only able to be retrieved in leinigen  when the lt terminates... so that's the reason for the lt launcher scripts.


Also, bash.exe is not returning the return code of the last run script, so windows passes the build as passed when it actually fails... Probably can use a small power shell script to throw an error if this occurs. 

Other then the above, I think it's ok as a initial commit for testing. There is a little bit of ugly stuff, but it can be cleaned up as we move along. 

In the future I want to move the the tests for appveyor to the "test:" section. I can use test2junit lein plugin to produce a junit file, but I have to figure out how to write some powerscript to upload all the xml files, could probably just write a shell command instead though...

kenny-evitt commented:
@justintaft I can't get the tests to run on one of my Windows 7 computers; output:

```
lein test lt.system-tests
OS is Windows 7
Exception in thread "main" org.openqa.selenium.SessionNotCreatedException: session not created exception
from unknown error: failed to close UI debuggers
  (Session info: content shell=)
  (Driver info: chromedriver=2.19.346078 (6f1f0cde889532d48ce8242342d0b84f94b114a1),platform=Windows NT 6.1 SP1 x86_64) (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 20.20 seconds
Build info: version: '2.47.0', revision: '0e4837e94b1fad0db93e88cd972ed3e235a22892', time: '2015-07-29 15:58:41'
System info: host: 'taugustine-PC', ip: '192.168.168.108', os.name: 'Windows 7', os.arch: 'amd64', os.version: '6.1', java.version: '1.8.0_40'
Driver info: org.openqa.selenium.remote.RemoteWebDriver
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
        at java.lang.reflect.Constructor.newInstance(Unknown Source)
        at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:206)
        at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:158)
        at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:595)
        at org.openqa.selenium.remote.RemoteWebDriver.startSession(RemoteWebDriver.java:242)
        at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:128)
        at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:155)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
        at java.lang.reflect.Constructor.newInstance(Unknown Source)
        at clojure.lang.Reflector.invokeConstructor(Reflector.java:180)
        at lt.webdriver_helper$connect_to_light_table_BANG_.invoke(webdriver_helper.clj:50)
        at lt.system_tests$lt_fixture.invoke(system_tests.clj:52)
        at clojure.test$compose_fixtures$fn__7139$fn__7140.invoke(test.clj:678)
        at clojure.test$default_fixture.invoke(test.clj:671)
        at clojure.test$compose_fixtures$fn__7139.invoke(test.clj:678)
        at clojure.test$test_all_vars$fn__7149.invoke(test.clj:717)
        at clojure.test$default_fixture.invoke(test.clj:671)
        at clojure.test$test_all_vars.invoke(test.clj:713)
        at clojure.test$test_ns.invoke(test.clj:736)
        at clojure.core$map$fn__4207.invoke(core.clj:2487)
        at clojure.lang.LazySeq.sval(LazySeq.java:42)
        at clojure.lang.LazySeq.seq(LazySeq.java:60)
        at clojure.lang.Cons.next(Cons.java:39)
        at clojure.lang.RT.boundedLength(RT.java:1654)
        at clojure.lang.RestFn.applyTo(RestFn.java:130)
        at clojure.core$apply.invoke(core.clj:619)
        at clojure.test$run_tests.doInvoke(test.clj:751)
        at clojure.lang.RestFn.applyTo(RestFn.java:137)
        at clojure.core$apply.invoke(core.clj:617)
        at user$eval85$fn__140$fn__171.invoke(form-init2747794511564858240.clj:1)
        at user$eval85$fn__140$fn__141.invoke(form-init2747794511564858240.clj:1)
        at user$eval85$fn__140.invoke(form-init2747794511564858240.clj:1)
        at user$eval85.invoke(form-init2747794511564858240.clj:1)
        at clojure.lang.Compiler.eval(Compiler.java:6619)
        at clojure.lang.Compiler.eval(Compiler.java:6609)
        at clojure.lang.Compiler.load(Compiler.java:7064)
        at clojure.lang.Compiler.loadFile(Compiler.java:7020)
        at clojure.main$load_script.invoke(main.clj:294)
        at clojure.main$init_opt.invoke(main.clj:299)
        at clojure.main$initialize.invoke(main.clj:327)
        at clojure.main$null_opt.invoke(main.clj:362)
        at clojure.main$main.doInvoke(main.clj:440)
        at clojure.lang.RestFn.invoke(RestFn.java:421)
        at clojure.lang.Var.invoke(Var.java:419)
        at clojure.lang.AFn.applyToHelper(AFn.java:163)
        at clojure.lang.Var.applyTo(Var.java:532)
        at clojure.main.main(main.java:37)
Tests failed.
Error encountered performing task 'test' with profile(s): 'dev'
Tests failed.
Return code is 1
===LIGHT TABLE DEBUG===
cat: /tmp/lt-debug: No such file or directory
Killing chrome
```

Weirdly (?), the return code seems to be being capture correctly.

 - [ ] We should also document or script Cygwin packages that need to be installed – *psmisc* (for `killall`), *unzip*, *wget*

justintaft commented:
@kenny-evitt did you use script/run-tests.sh ?

kenny-evitt commented:
@justintaft Yeah

justintaft commented:
@kenny-evitt hmmmm, i've ran into scenarios where It would not connect. I kill light table, java, and chrome driver, and then retry a few times. Eventually it works. It's weird how it always seems to work on travis-ci and appveyor...

I've seen light table not finish initialize (ie, stuck on the light-table into screen and does not accept input).  Haven't had the time to dig into why this happens. This happens sporadically when it's trying to connect to chrome driver... 

This might be able to be fixed if we can modify light table to accept a chromedriver debug port to connect to on init rather then being hard coded...

kenny-evitt commented:
@justintaft I got another error just now; the same one each of the three times I ran `script/run-tests.sh`:

```
$ script/run-tests.sh
Checking for chromedrive
Killing all chromedrivers & starting new instance.
chromedriver: no process found
chromedriver.exe: no process found
Running tests
Starting ChromeDriver 2.19.346078 (6f1f0cde889532d48ce8242342d0b84f94b114a1) on port 9515
Only local connections are allowed.
Retrieving lein-cljsbuild/cljs-compat/1.0.0-SNAPSHOT/cljs-compat-1.0.0-20151101.091649-38.pom from clojars
Retrieving lein-cljsbuild/cljs-compat/1.0.0-SNAPSHOT/cljs-compat-1.0.0-20151101.091649-38.jar from clojars

lein test lt.system-tests
OS is Windows 7
Exception in thread "main" org.openqa.selenium.WebDriverException: unknown error: cannot connect to chrome at 127.0.0.1:8315
from chrome not reachable
  (Driver info: chromedriver=2.19.346078 (6f1f0cde889532d48ce8242342d0b84f94b114a1),platform=Windows NT 6.1 SP1 x86_64) (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 60.28 seconds
Build info: version: '2.47.0', revision: '0e4837e94b1fad0db93e88cd972ed3e235a22892', time: '2015-07-29 15:58:41'
System info: host: 'not-my-real-host-name', ip: '192.168.168.108', os.name: 'Windows 7', os.arch: 'amd64', os.version: '6.1', java.version: '1.8.0_40'
Driver info: org.openqa.selenium.remote.RemoteWebDriver
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
        at java.lang.reflect.Constructor.newInstance(Unknown Source)
        at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:206)
        at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:158)
        at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:595)
        at org.openqa.selenium.remote.RemoteWebDriver.startSession(RemoteWebDriver.java:242)
        at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:128)
        at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:155)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
        at java.lang.reflect.Constructor.newInstance(Unknown Source)
        at clojure.lang.Reflector.invokeConstructor(Reflector.java:180)
        at lt.webdriver_helper$connect_to_light_table_BANG_.invoke(webdriver_helper.clj:50)
        at lt.system_tests$lt_fixture.invoke(system_tests.clj:52)
        at clojure.test$compose_fixtures$fn__7139$fn__7140.invoke(test.clj:678)
        at clojure.test$default_fixture.invoke(test.clj:671)
        at clojure.test$compose_fixtures$fn__7139.invoke(test.clj:678)
        at clojure.test$test_all_vars$fn__7149.invoke(test.clj:717)
        at clojure.test$default_fixture.invoke(test.clj:671)
        at clojure.test$test_all_vars.invoke(test.clj:713)
        at clojure.test$test_ns.invoke(test.clj:736)
        at clojure.core$map$fn__4207.invoke(core.clj:2487)
        at clojure.lang.LazySeq.sval(LazySeq.java:42)
        at clojure.lang.LazySeq.seq(LazySeq.java:60)
        at clojure.lang.Cons.next(Cons.java:39)
        at clojure.lang.RT.boundedLength(RT.java:1654)
        at clojure.lang.RestFn.applyTo(RestFn.java:130)
        at clojure.core$apply.invoke(core.clj:619)
        at clojure.test$run_tests.doInvoke(test.clj:751)
        at clojure.lang.RestFn.applyTo(RestFn.java:137)
        at clojure.core$apply.invoke(core.clj:617)
        at user$eval85$fn__140$fn__171.invoke(form-init3575876859897388307.clj:1)
        at user$eval85$fn__140$fn__141.invoke(form-init3575876859897388307.clj:1)
        at user$eval85$fn__140.invoke(form-init3575876859897388307.clj:1)
        at user$eval85.invoke(form-init3575876859897388307.clj:1)
        at clojure.lang.Compiler.eval(Compiler.java:6619)
        at clojure.lang.Compiler.eval(Compiler.java:6609)
        at clojure.lang.Compiler.load(Compiler.java:7064)
        at clojure.lang.Compiler.loadFile(Compiler.java:7020)
        at clojure.main$load_script.invoke(main.clj:294)
        at clojure.main$init_opt.invoke(main.clj:299)
        at clojure.main$initialize.invoke(main.clj:327)
        at clojure.main$null_opt.invoke(main.clj:362)
        at clojure.main$main.doInvoke(main.clj:440)
        at clojure.lang.RestFn.invoke(RestFn.java:421)
        at clojure.lang.Var.invoke(Var.java:419)
        at clojure.lang.AFn.applyToHelper(AFn.java:163)
        at clojure.lang.Var.applyTo(Var.java:532)
        at clojure.main.main(main.java:37)
Tests failed.
Error encountered performing task 'test' with profile(s): 'dev'
Tests failed.
Return code is 1
===LIGHT TABLE DEBUG===
cat: /tmp/lt-debug: No such file or directory
Killing chrome
```

justintaft commented:
@kenny-evitt  Ok, i*think* the issue is a incorrect path to cygwin's bash.exe . The script assumes c:\cygwin\... is the base directory. I assume you do not see light table start? Does cygwin automatically add it's bin directory to windows PATH? Ie, can you type "bash.exe" in console to start it from any directory, outside of cygwin?

https://github.com/LightTable/LightTable/pull/1996/files#diff-684e38ea5f35918a039a2474df275b7cR3

I was able to find the above out because /tmp/lt-debug should be created when running light table with the scripts. Since your log said the file was not found, this was a bit suspicious...

kenny-evitt commented:
@justintaft Nice work! :tada:

justintaft commented:
@kenny-evitt thanks =),  so it's working on your computer now?

kenny-evitt commented:
@justintaft Still no luck:

```
$ script/run-tests.sh
Checking for chromedrive
Downloading & extracting chromedriver.
--2015-11-15 16:46:36--  http://chromedriver.storage.googleapis.com/2.19/chromedriver_win32.zip
Resolving chromedriver.storage.googleapis.com (chromedriver.storage.googleapis.com)... 108.177.11.128, 2607:f8b0:400c:c06::80
Connecting to chromedriver.storage.googleapis.com (chromedriver.storage.googleapis.com)|108.177.11.128|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2468633 (2.4M) [application/x-zip-compressed]
Saving to: ‘/tmp/chromedriver.zip’

/tmp/chromedriver.z 100%[=====================>]   2.35M  1.93MB/s   in 1.2s

2015-11-15 16:46:37 (1.93 MB/s) - ‘/tmp/chromedriver.zip’ saved [2468633/2468633]

Archive:  /tmp/chromedriver.zip
  inflating: /tmp/chromedriver.exe
Killing all chromedrivers & starting new instance.
script/run-tests.sh: line 40: killall: command not found
script/run-tests.sh: line 41: killall: command not found
Running tests
Starting ChromeDriver 2.19.346078 (6f1f0cde889532d48ce8242342d0b84f94b114a1) on port 9515
Only local connections are allowed.

lein test lt.system-tests
Exception in thread "main" org.openqa.selenium.SessionNotCreatedException: session not created exception
from unknown error: failed to close UI debuggers
  (Session info: content shell=)
  (Driver info: chromedriver=2.19.346078 (6f1f0cde889532d48ce8242342d0b84f94b114a1),platform=Windows NT 6.1 SP1 x86) (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 27.23 seconds
Build info: version: '2.47.0', revision: '0e4837e94b1fad0db93e88cd972ed3e235a22892', time: '2015-07-29 15:58:41'
System info: host: 'Kelvin', ip: '192.168.1.59', os.name: 'Windows 7', os.arch: 'x86', os.version: '6.1', java.version: '1.7.0_71'
Driver info: org.openqa.selenium.remote.RemoteWebDriver
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
        at java.lang.reflect.Constructor.newInstance(Unknown Source)
        at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:206)
        at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:158)
        at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:595)
        at org.openqa.selenium.remote.RemoteWebDriver.startSession(RemoteWebDriver.java:242)
        at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:128)
        at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:155)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
        at java.lang.reflect.Constructor.newInstance(Unknown Source)
        at clojure.lang.Reflector.invokeConstructor(Reflector.java:180)
        at lt.webdriver_helper$connect_to_light_table_BANG_.invoke(webdriver_helper.clj:46)
        at lt.system_tests$lt_fixture.invoke(system_tests.clj:52)
        at clojure.test$compose_fixtures$fn__7139$fn__7140.invoke(test.clj:678)
        at clojure.test$default_fixture.invoke(test.clj:671)
        at clojure.test$compose_fixtures$fn__7139.invoke(test.clj:678)
        at clojure.test$test_all_vars$fn__7149.invoke(test.clj:717)
        at clojure.test$default_fixture.invoke(test.clj:671)
        at clojure.test$test_all_vars.invoke(test.clj:713)
        at clojure.test$test_ns.invoke(test.clj:736)
        at clojure.core$map$fn__4207.invoke(core.clj:2487)
        at clojure.lang.LazySeq.sval(LazySeq.java:42)
        at clojure.lang.LazySeq.seq(LazySeq.java:60)
        at clojure.lang.Cons.next(Cons.java:39)
        at clojure.lang.RT.boundedLength(RT.java:1654)
        at clojure.lang.RestFn.applyTo(RestFn.java:130)
        at clojure.core$apply.invoke(core.clj:619)
        at clojure.test$run_tests.doInvoke(test.clj:751)
        at clojure.lang.RestFn.applyTo(RestFn.java:137)
        at clojure.core$apply.invoke(core.clj:617)
        at user$eval85$fn__140$fn__171.invoke(form-init9018621538281178546.clj:1)
        at user$eval85$fn__140$fn__141.invoke(form-init9018621538281178546.clj:1)
        at user$eval85$fn__140.invoke(form-init9018621538281178546.clj:1)
        at user$eval85.invoke(form-init9018621538281178546.clj:1)
        at clojure.lang.Compiler.eval(Compiler.java:6619)
        at clojure.lang.Compiler.eval(Compiler.java:6609)
        at clojure.lang.Compiler.load(Compiler.java:7064)
        at clojure.lang.Compiler.loadFile(Compiler.java:7020)
        at clojure.main$load_script.invoke(main.clj:294)
        at clojure.main$init_opt.invoke(main.clj:299)
        at clojure.main$initialize.invoke(main.clj:327)
        at clojure.main$null_opt.invoke(main.clj:362)
        at clojure.main$main.doInvoke(main.clj:440)
        at clojure.lang.RestFn.invoke(RestFn.java:421)
        at clojure.lang.Var.invoke(Var.java:419)
        at clojure.lang.AFn.applyToHelper(AFn.java:163)
        at clojure.lang.Var.applyTo(Var.java:532)
        at clojure.main.main(main.java:37)
Tests failed.
Error encountered performing task 'test' with profile(s): 'dev'
Tests failed.
Return code is 1
===LIGHT TABLE DEBUG===
./script/light.sh: line 22: /cygdrive/c/@code/LightTable/deploy/electron/electron/electron.exe: No such file or directory
Killing chrome
```

kenny-evitt commented:
LT was opened when I ran *run-tests.sh* just now tho maybe it still didn't work:

```
$ script/run-tests.sh
Checking for chromedrive
Killing all chromedrivers & starting new instance.
chromedriver: no process found
chromedriver.exe: no process found
Running tests
Starting ChromeDriver 2.19.346078 (6f1f0cde889532d48ce8242342d0b84f94b114a1) on port 9515
Only local connections are allowed.

lein test lt.system-tests

lein test :only lt.system-tests/new-file

ERROR in (new-file) (util.clj:337)
Uncaught exception, not in assertion.
expected: nil
  actual: org.openqa.selenium.NoSuchElementException:
When an element cannot be found in clj-webdriver, nil is returned. You've just tried to perform an action on an element that returned as nil for the search query you used. Please verify the query used to locate this element; it is not on the current page.
For documentation on this error, please visit: http://seleniumhq.org/exceptions/no_such_element.html
Build info: version: '2.47.0', revision: '0e4837e94b1fad0db93e88cd972ed3e235a22892', time: '2015-07-29 15:58:41'
System info: host: 'Kelvin', ip: '192.168.1.59', os.name: 'Windows 7', os.arch: 'x86', os.version: '6.1', java.version: '1.7.0_71'
Driver info: driver.version: unknown
 at clj_webdriver.util$throw_nse.invoke (util.clj:337)
    clj_webdriver.util$throw_nse.invoke (util.clj:335)
    clj_webdriver.core$eval3451$fn__3454.invoke (core_element.clj:433)
    clj_webdriver.core$eval2589$fn__2645$G__2566__2652.invoke (core.clj:115)
    clj_webdriver.taxi$send_keys.invoke (taxi.clj:1228)
    clj_webdriver.taxi$send_keys.invoke (taxi.clj:1226)
    lt.system_tests$helper_select_files.invoke (system_tests.clj:45)
    lt.system_tests/fn (system_tests.clj:83)
    clojure.test$test_var$fn__7145.invoke (test.clj:701)
    clojure.test$test_var.invoke (test.clj:701)
    clojure.test$test_all_vars$fn__7149$fn__7156.invoke (test.clj:717)
    lt.system_tests$lt_fixture.invoke (system_tests.clj:53)
    clojure.test$compose_fixtures$fn__7139$fn__7140.invoke (test.clj:678)
    clojure.test$default_fixture.invoke (test.clj:671)
    clojure.test$compose_fixtures$fn__7139.invoke (test.clj:678)
    clojure.test$test_all_vars$fn__7149.invoke (test.clj:717)
    clojure.test$default_fixture.invoke (test.clj:671)
    clojure.test$test_all_vars.invoke (test.clj:713)
    clojure.test$test_ns.invoke (test.clj:736)
    clojure.core$map$fn__4207.invoke (core.clj:2487)
    clojure.lang.LazySeq.sval (LazySeq.java:42)
    clojure.lang.LazySeq.seq (LazySeq.java:60)
    clojure.lang.Cons.next (Cons.java:39)
    clojure.lang.RT.boundedLength (RT.java:1654)
    clojure.lang.RestFn.applyTo (RestFn.java:130)
    clojure.core$apply.invoke (core.clj:619)
    clojure.test$run_tests.doInvoke (test.clj:751)
    clojure.lang.RestFn.applyTo (RestFn.java:137)
    clojure.core$apply.invoke (core.clj:617)
    user$eval85$fn__140$fn__171.invoke (form-init7485580914009611721.clj:1)
    user$eval85$fn__140$fn__141.invoke (form-init7485580914009611721.clj:1)
    user$eval85$fn__140.invoke (form-init7485580914009611721.clj:1)
    user$eval85.invoke (form-init7485580914009611721.clj:1)
    clojure.lang.Compiler.eval (Compiler.java:6619)
    clojure.lang.Compiler.eval (Compiler.java:6609)
    clojure.lang.Compiler.load (Compiler.java:7064)
    clojure.lang.Compiler.loadFile (Compiler.java:7020)
    clojure.main$load_script.invoke (main.clj:294)
    clojure.main$init_opt.invoke (main.clj:299)
    clojure.main$initialize.invoke (main.clj:327)
    clojure.main$null_opt.invoke (main.clj:362)
    clojure.main$main.doInvoke (main.clj:440)
    clojure.lang.RestFn.invoke (RestFn.java:421)
    clojure.lang.Var.invoke (Var.java:419)
    clojure.lang.AFn.applyToHelper (AFn.java:163)
    clojure.lang.Var.applyTo (Var.java:532)
    clojure.main.main (main.java:37)

lein test lt.webdriver-helper

Ran 1 tests containing 1 assertions.
0 failures, 1 errors.
Tests failed.
Error encountered performing task 'test' with profile(s): 'dev'
Tests failed.
0 failures, 1 errors.
Return code is 0
===LIGHT TABLE DEBUG===
Killing chrome
```

I think it failed ('errored') because of the Windows Firewall non-access. I re-ran it and everything seemed to work (i.e. fail) as expected:

```
$ script/run-tests.sh
Checking for chromedrive
Killing all chromedrivers & starting new instance.
chromedriver: no process found
chromedriver.exe: no process found
Running tests
Starting ChromeDriver 2.19.346078 (6f1f0cde889532d48ce8242342d0b84f94b114a1) on port 9515
Only local connections are allowed.

lein test lt.system-tests

lein test :only lt.system-tests/new-file

FAIL in (new-file) (system_tests.clj:107)
expected: (= "File Persistence Test I SHOULD FAIL" (taxi/text {:css ".CodeMirror-code span"}))
  actual: (not (= "File Persistence Test I SHOULD FAIL" "File Persistence Test"))

lein test lt.webdriver-helper

Ran 1 tests containing 3 assertions.
1 failures, 0 errors.
Tests failed.
Error encountered performing task 'test' with profile(s): 'dev'
Tests failed.
Return code is 1
===LIGHT TABLE DEBUG===
Killing chrome
```

LT wasn't closed tho. Is that expected?

justintaft commented:
Bleh, I still have to handle that scenario, thanks.

justintaft commented:
@kenny-evitt alright, the LT launched by the test is now killed at the end of tests. I needed to use a couple "exec"s in a few bash scripts so the kill signal would propagate correctly. In the future a few flags can be added so LT isn't auto killed. This will allow us to write selenium tests in a REPL much easier.

The PR should be good for review. 

Like I've mentioned, sometimes while testing LT does not initialize all the way.... not sure what causes this and will need a little investigation.

kenny-evitt commented:
@justintaft The AppVeyor failure isn't strictly an npm package download. It's a package that's trying to download Electron from its GitHub release and GitHub is rate limiting requests. According to the warning text we might be able to bypass the limit by authenticating; with a SSH key or, probably better, by generating an OAuth token in GitHub, encrypting it in AppVeyor, and then adding the encrypted token to the *appveyor.yml*. Beyond that, I'm not sure where we could cache the Electron release files.

I'm going to test the tests on one of my Ubuntu boxes now.

kenny-evitt commented:
@justintaft The tests ran but failed – the full text wasn't added to the files. Here's the console output from running *script/run-tests.sh*:

```
Checking for chromedrive...
Downloading & extracting chromedriver.
--2015-11-18 20:38:34--  http://chromedriver.storage.googleapis.com/2.19/chromedriver_linux64.zip
Resolving chromedriver.storage.googleapis.com (chromedriver.storage.googleapis.com)... 173.194.214.128, 2607:f8b0:400c:c0b::80
Connecting to chromedriver.storage.googleapis.com (chromedriver.storage.googleapis.com)|173.194.214.128|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2540107 (2.4M) [application/zip]
Saving to: ‘/tmp/chromedriver.zip’

100%[======================================>] 2,540,107   6.79MB/s   in 0.4s   

2015-11-18 20:38:35 (6.79 MB/s) - ‘/tmp/chromedriver.zip’ saved [2540107/2540107]

Archive:  /tmp/chromedriver.zip
  inflating: /tmp/chromedriver       
Killing all chromedrivers & starting new instance.
Running tests
Starting ChromeDriver 2.19.346067 (6abd8652f8bc7a1d825962003ac88ec6a37a82f1) on port 9515
Only local connections are allowed.

lein test lt.system-tests

lein test :only lt.system-tests/new-file

FAIL in (new-file) (system_tests.clj:88)
expected: (= "File Persistence Test" (read-file-contents temp-file-path))
  actual: (not (= "File Persistence Test" "stence Test"))

lein test :only lt.system-tests/new-file

FAIL in (new-file) (system_tests.clj:105)
expected: (= "File Persistence Test" (taxi/text {:css ".CodeMirror-code span"}))
  actual: (not (= "File Persistence Test" "stence Test"))

lein test lt.webdriver-helper

Ran 1 tests containing 2 assertions.
2 failures, 0 errors.
Tests failed.
Error encountered performing task 'test' with profile(s): 'dev'
Tests failed.
===LIGHT TABLE DEBUG===
[6559:1118/203856:ERROR:browser_main_loop.cc(185)] Running without the SUID sandbox! See https://code.google.com/p/chromium/wiki/LinuxSUIDSandboxDevelopment for more information on developing with the sandbox on.
[6646:1118/203856:ERROR:sandbox_linux.cc(345)] InitializeSandbox() called with multiple threads in process gpu-process
[6658:1118/203857:INFO:renderer_main.cc(200)] Renderer process started
[6559:1118/203902:WARNING:accelerator_util.cc(185)] Invalid accelerator token: x control
[6559:1118/203902:WARNING:accelerator_util.cc(185)] Invalid accelerator token: x k
[6559:1118/203902:WARNING:accelerator_util.cc(185)] Invalid accelerator token: x control
[6559:1118/203902:WARNING:accelerator_util.cc(185)] Invalid accelerator token: x k
Killing chrome
```

justintaft commented:
@kenny-evitt i think a VIM plugin is loading, causing them to fail ;). 

```
FAIL in (new-file) (system_tests.clj:105)
expected: (= "File Persistence Test" (taxi/text {:css ".CodeMirror-code span"}))
  actual: (not (= "File Persistence Test" "stence Test"))
```

`Fi` tries to find the character I behind the cursor, a bunch of vim commands run, then the next 'i' is hit, which causes insertion mode to happen, and "stence Test" is then typed. Thats my guess though =).

justintaft commented:
@kenny-evitt can you provide an oath token? Maybe this should be tied to the LightTable account?

kenny-evitt commented:
@justintaft You're probably right about the VIM plugin.

I'll add the encrypted token to the *appveyor.yml* file. I'll generate a token for the organization.

kenny-evitt commented:
@justintaft We need a way to ignore, or (temporarily) uninstall or disable, non-default plugins. And maybe reset behaviors and keymap files too.

justintaft commented:
@kenny-evitt i create a new "home" directory for the test, so any user configuration files won't load. If it should load during the test, they should add it to their test =).

I'm going to be busy for the next month, won't have much time to code. I didn't have the time to look into how to use the auth token. 

I'll be around if you have questions!

justintaft commented:
@kenny-evitt @cldwalker any thoughts? should be good for review =)

cldwalker commented:
@justintaft Thanks for your work on this! Will look into this after 0.8.0 is released. I'm seeing unrelated commits in this branch e.g. 220f3995b107687082e70f7331ea7dc9599c2567. Probably best to just a create a fresh branch that cherry-picks your commits

kenny-evitt commented:
@justintaft When you have a chance, rebase the latest upstream *master* onto your topic branch.

After the 0.8.0 release, we can schedule a time to chat about next steps for this.

justintaft commented:
@kenny-evitt @cldwalker rebased. 

Tests will fail randomly. This is because  'require("clipboard")' fails, an exception is thrown and LT does not continue booting.  If you re-run  script/run-test.sh multiple times, the tests will passed. 

Not sure why this happens, only found https://github.com/atom/electron/issues/916

justintaft commented:
Hm, digged a little deeper.

When using 'require',  Modules_resolveLookupPaths is used. A Parent object is passed to this function. 
When LT is boots up normally, the parent file name is "LightTable.html". When LT Fails to boot up, the parent file name is "init.js" ....

justintaft commented:
Problem is resources can't be accessed in atom.asar sometimes... I have the right permissions, the file 
exists.   The below happens when LT doesn't boot up.  When LT boots up, it get a listing of files in the asar archive...
```
fs.readdirSync("/Users/justintaft/programming/LightTableElectron/LightTable/deploy/electron/electron/Electron.app/Contents/Resources/atom.asar");

Uncaught Error: ENOENT,  not found in /Users/justintaft/programming/LightTableElectron/LightTable/deploy/electron/electron/Electron.app/Contents/Resources/atom.asar(…)```

justintaft commented:
Looks like bug is known, waiting for 
atom/electron#3699 
to be fixed

kenny-evitt commented:
@justintaft What about switching to using a `file://` URL? The original comment in atom/electron#3699 mentions that that works.

justintaft commented:
We are using file:/// already  =/

kenny-evitt commented:
@justintaft That's annoying. No other workarounds mentioned in the Electron issue?

justintaft commented:
@kenny-evitt looks like a fix was merged into electron =) https://github.com/atom/electron/issues/3699#issuecomment-168056601

I'll check this out in a few days and see if it works

Commit:
Updates electron to 0.36.7

justintaft commented:
@kenny-evitt Cool, it looks like that fixed our issue =)

The windows build is failing, will check this out soon.

justintaft commented:
@kenny-evitt also, did you generate an auth token?

I wonder how we can protect the token and still allow pull requests to trigger builds...
https://www.appveyor.com/docs/build-configuration#secure-variables
https://github.com/atom/grunt-download-electron

kenny-evitt commented:
@justintaft I found [the old Gitter comment](https://gitter.im/LightTable/LightTable?at=5650c32ca051fea34242f56d). Here's the encrypted token:

```
Dm2epqg+b/o1WSR6e+hYFe1O12Vr10yshOFFJ3ISqke8Xl9wl1JB73B7wa4oYn4q
```

Only the AppVeyor account can decrypt it so it's safe to include in the *appveyor.yml* file.

kenny-evitt commented:
Just for anyone wondering what that encrypted token *is* (like myself in any more than a few days), that's a Personal Access Token for my GitHub account with [the *public_repo* scope](https://developer.github.com/v3/oauth/#scopes):

> Grants read/write access to code, commit statuses, collaborators, and deployment statuses for public repositories and organizations. Also required for starring public repositories.

@justintaft Why did we want write/push access from AppVeyor? For uploading build artifacts?

justintaft commented:
@kenny-evitt grunt-electrond-download needed to perform some github requests for downloading.

https://ci.appveyor.com/project/kenny-evitt/lighttable/build/1.0.114
```
>> Cannot find electron v0.36.7 from GitHub [Error: API rate limit exceeded for 104.197.110.30. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)]
Warning: Task "download-electron" failed. Use --force to continue.
```

I'm not sure if we needed write and push access....
Also, the previous appveyor page I link mentioned encrypted tokens won't decrypt for pull requests. This is to prevent people creating a malicious script to grab the unencrypted token,

justintaft commented:
@cldwalker @kenny-evitt Electron has been updated in this PR. They pushed a required fix that was causing the application not to initialize properly when webdriver was attached.

Kenny has created a new account, light-table-bot, with a public read-only oauth token to prevent api rate limits during the build process.

The following is left to do:
- Fix race conditions in the tests. Appveyor is slow, so I have to add timeouts/checks in various places to wait for elements to be clickable.
- Enable OSX support for travis (this can probably be in a separate build)
- Clean up code
-Add documentation

This will be a couple weekends worth of work.

After the PR is merged, maybe we can have a little party in SF? I just moved here and it would be cool to meet the team =).

kenny-evitt commented:
@justintaft Great work! Sadly, none of the current team is in the Bay area so we can only tele-party. You could reach out to the original Kodowa team but they're almost certainly busy with their new project [Eve](http://incidentalcomplexity.com/).

Mouvedia commented:
> The following is left to do:

> 1. Fix race conditions in the tests. Appveyor is slow, so I have to add timeouts/checks in various places to wait for elements to be clickable.
2. Enable OSX support for travis (this can probably be in a separate PR)
3. Clean up code
4. Add documentation

@justintaft Can you at least take care of 1. and 3. so that we can merge this PR? Once that's done you should probably open a new task issue which will list 2. and 4. in a GFM task list.

justintaft commented:
@Mouvedia I'll take a look during the next 2 weeks. I've been a bit busy.

justintaft commented:
@kenny-evitt  can you re-enable appveyor?

kenny-evitt commented:
@justintaft It seems like it wasn't disabled.

Note that I'm not helping out with Light Table currently so, in general, reach out to those that are first. I still get explicit mentions tho and I'll try to respond in at least a somewhat timely fashion.

Mouvedia commented on .travis.yml:
```diff
@@ -3,8 +3,9 @@ env:
  - linux
- - osx
-lein: clojure
+# Clojure is not supported on osx
```
https://docs.travis-ci.com/user/languages/clojure/#What-This-Guide-Covers

justintaft replied:
The page says "Clojure builds are not available on the OSX environment." Maybe lein and be fetched separately?

justintaft commented:
@cldwalker @Mouvedia  Good to review. 

This does NOT contain OSX support on travis. If there are necessary changes that are required for the PR and they are minimal, I will complete them quickly. 

Please note I am fairly busy, and I don't have extra time to contribute to Light Table. I can respond to a questions though if someone wants to further working on the test suite.

Currently Light Table is opened, a new file is created and saved, then contents are checked.

CI links are 
https://ci.appveyor.com/project/kenny-evitt/lighttable/history
https://travis-ci.org/LightTable/LightTable/pull_requests

Appveyor has to be added to the github project (it's not being used in the "all checks have passed" section listed above).

Mouvedia commented:
Most* of the bullets of the [QA](../wiki/QA-Checklist) should be converted to checkboxes so that we can tick the ones that you have already covered.

\* not everything can be tested […]

justintaft commented:
#1 is taken care of. I implemented a polling function which re-runs until it passes (ie, waiting for an element to become visible). Default is wait 1 second at most, and re-try every 100 ms.

#3, I cleaned up the code a bit but it honestly can be cleaned up more. I just don't have the time to in the foreseeable future, sorry.

I tackled #4 a bit, gave a high level overview in docs/testing.md and added docstrings to functions.

justintaft commented:
OSX is now being tested too.

Updated .travis.yml file to manually install leiningen, worked out of the box. 
Hopefully this PR will save the LT team a bunch of time =).

Mouvedia commented on .travis.yml:
```diff
@@ -1,11 +1,26 @@
-  - mrundberget@hotmail.com
+    - gabriel.horner@gmail.com
+    - kenny.evitt@gmail.com
+    - mrundberget@hotmail.com
```
You could add `sbauer322@gmail.com` as well.
cf https://github.com/orgs/LightTable/people

Mouvedia commented:
@lighttable/maintainers can one of you review this PR?

sbauer322 commented:
Due to the large number of changes, and my lack of familiarity with the topic, I will be deferring to the other maintainers for this review.

Mouvedia commented on test/lt/webdriver_helper.clj:
```diff
@@ -0,0 +1,121 @@
+
+;Defines temporary directory for testing light table
+(def temp-home-directory
+    (str (Files/createTempDirectory "lightable_homedir_for_testing" (into-array FileAttribute []))))
```
Should this be `lighttable_homedir_for_testing`?

Mouvedia commented on test/lt/lt-test-plugin/src/lt/plugins/lighttable-test-plugin.cljs:
```diff
@@ -0,0 +1,34 @@
+(ns lt.plugins.lighttable-test-plugin
+ "Mocks & extends light table components for auotmated testing."
```
automated

Mouvedia commented on test/lt/lt-test-plugin/src/lt/plugins/lighttable-test-plugin.cljs:
```diff
@@ -0,0 +1,34 @@
+(defn show-file-dialog-mock
+  "Mocks show-file-dialog so native file dialog does not show.
+   A popup dialog with a text field is shown instead. File paths
+   can be specified in the path fields, seperated by commas to donate
```
separated

justintaft commented:
Is there a need for system-level tests on macOS, Linux, and windows? I can update the PR if needed.

prertik commented:
@justintaft sure! if you have the time. This looks good, by my changes(in local) need to be somewhat modified after looking at this.
